db/db.go

```diff
@@ -28,16 +28,36 @@ type DB interface {
 	Impl() any
 }
 
-// Entry is a database entry consisting of a Key and Value
-type Entry struct {
-	Key   []byte
-	Value []byte
+// Iterator is an iterator over a DB's key/value pairs.
+type Iterator interface {
+	// Valid returns true if the iterator is positioned at a valid key/value pair.
+	Valid() bool
+
+	// Next moves the iterator to the next key/value pair. It returns whether the
+	// iterator is valid after the call. Once invalid, the iterator remains
+	// invalid.
+	Next() bool
+
+	// Key returns the key at the current position.
+	Key() []byte
+
+	// Value returns the value at the current position.
+	Value() []byte
+
+	// Close closes the iterator.
+	Close() error
+
+	// Seek would seek to the provided key if present. If absent, it would seek to the next
+	// key in lexicographical order
+	Seek(key []byte) bool
 }
 
 // Transaction provides an interface to access the database's state at the point the transaction was created
 // Updates done to the database with a transaction should be only visible to other newly created transaction after
 // the transaction is committed.
 type Transaction interface {
+	// NewIterator returns an iterator over the database's key/value pairs.
+	NewIterator() (Iterator, error)
```

Comment:
Shouldn't this be defined on db interface instead of the transaction?

Reply:
Nah. The iterations are made on a transaction

Reply:
Database is volatile, can change any time. Transactions represent a fixed point in database's history. So, iterators make more sense on a fixed view of the database rather than a one that is possibly changing continuously .